crates/shim-sgx/src/handler/mod.rs

```diff
@@ -815,37 +956,48 @@ impl<'a> Handler<'a> {
 
         // Process the ledger first, before doing anything else, because it can
         // legitly fail when running out of resources.
-        if let Err(e) = heap.munmap(addr, length) {
-            debugln!(self, "[{tid}] ERROR munmap: heap.munmap FAILED !!! {e:?}");
-            return Err(ENOMEM);
-        }
+        heap.munmap_with(addr, length, move |record: &Record<Access>| {
+            debugln!(self, "[{tid}] sgx_unmap({:#?}", record);
 
-        // On the other hand, failing in any of these operations is expected to
-        // crash the enclave because it is due either to a software bug, or a
-        // malicious host.
-        if let Err(e) = self.modify_sgx_page_type(addr_in, length.bytes(), Class::Trimmed as _) {
-            debugln!(
-                self,
-                "[{tid}] ERROR munmap: modify_sgx_page_type FAILED !!! {e:?}"
-            );
-            self.attacked();
-        }
+            if !record.access.contains(Access::MMAPPED) {
+                // if not MMAPPED, we don't have to do anything
+                return;
+            }
 
-        for i in 0..pages {
-            let virt_addr = VirtAddr::new((addr.raw() + i * Page::SIZE) as u64);
-            // # Safety
-            //
-            // The address must be page aligned.
-            let page_addr = unsafe { PageAddr::from_start_address_unchecked(virt_addr) };
+            let region = &record.region;
+            let addr = NonNull::new(region.start.as_mut_ptr() as *mut c_void).unwrap();
```

Comment:
Should `Region` have `addr()`, `len()`, and `pages()`, methods? Or maybe `Address` could benefit from a built-in conversion to `NonNull`, since this also happens several times below.

Reply:
added